processing/src/main/java/io/druid/query/groupby/having/DimFilterHavingSpec.java

```diff
@@ -25,26 +25,29 @@
 import io.druid.data.input.Row;
 import io.druid.query.filter.DimFilter;
 import io.druid.query.filter.ValueMatcher;
-import io.druid.query.groupby.RowBasedValueMatcherFactory;
+import io.druid.query.groupby.RowBasedColumnSelectorFactory;
+import io.druid.segment.column.ValueType;
 
 import java.nio.ByteBuffer;
+import java.util.Map;
 
-public class DimFilterHavingSpec implements HavingSpec
+public class DimFilterHavingSpec extends BaseHavingSpec
 {
   private static final byte CACHE_KEY = (byte) 0x9;
 
   private final DimFilter dimFilter;
-  private final RowBasedValueMatcherFactory valueMatcherFactory;
-  private final ValueMatcher valueMatcher;
+  private final ThreadLocal<Row> rowSupplier;
+
+  private final Object valueMatcherMonitor = new Object();
+  private volatile ValueMatcher valueMatcher;
```

Comment:
Are you talking about the risk that a ValueMatcher impl will not be safely initialized and that its internal state will not be properly visible to another thread? In that case, yeah, that's a good point. I don't think any current ValueMatchers would have this behavior but I suppose they might in theory.

Or are you saying that it's possible for an object that was safely initialized, and written to a volatile reference, to be then read from another thread as a null reference? In my understanding of the JMM, that is not possible, since writing a safely initialized object to a volatile reference will cause the reference and also the object's initialized state to be visible to any thread that reads the reference.

Either way, I agree we can get rid of thread safetiness in DimFilterHavingSpec so this point is moot as far as the PR goes.